### PR TITLE
TriBITS update to allow usage in Trilinos auto PR file change logic (#3133)

### PR DESCRIPTION
**Pull request by bartlettroscoe**
@william76, @william76, @fryeguy52 

## Description

This PR branch contains a TriBITS snapshot update to allow the Trilinos auto PR tester to use TriBITS logic in determining what to enable based on `git diff --name-only <branch-tip> ^github/devleop`.  (See #3133).  This includes specialized logic for what file changes in Trilinos should trigger a global build.

I added strong strong automated testing for this new script and added a new TriBITS package called `TrilinosFrameworkTests` to run these automated tests automatically.

## Motivation and Context

See the motivation in #3133.

## How Has This Been Tested?

In addition to the strong automated tests added in the new `TrilinosFrameworkTests` package, I did some manual testing.  For example, I looked at a large set of changes between the tip of this branch and a version of Trilinos 'develop' from 7/6/2018 de924c.

```
 time ~/Trilinos.base/Trilinos/commonTools/framework/get-changed-trilinos-packages.sh de924c 3133-tribits-file-change-logic enablePackages.cmake

***
*** Generating set of Trilinos enables given modified packages from
**** git commit de924c to 3133-tribits-file-change-logic
***

TRILINOS_DIR=/home/rabartl/Trilinos.base/Trilinos/commonTools/framework/../..
TRIBITS_DIR=/home/rabartl/Trilinos.base/Trilinos/commonTools/framework/../../cmake/tribits

A) Generate the TrilinosPackageDependencies.xml file


B) Get the set of changed files


Current directory: /home/rabartl/Trilinos.base/Trilinos

Writing file /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/MPI_RELEASE_DEBUG_SHARED/changed-files.txt ...

git diff --name-only de924c..3133-tribits-file-change-logic > /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/MPI_RELEASE_DEBUG_SHARED/changed-files.txt

Current directory: /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/MPI_RELEASE_DEBUG_SHARED

C) Get the unfiltered list of changed Trilinos packages (including 'ALL_PACKAGES')

CHANGED_PACKAGES_FULL_LIST='ALL_PACKAGES,TrilinosFrameworkTests,Amesos2,Anasazi,Belos,Galeri,Ifpack2,Kokkos,MueLu,PanzerAdaptersSTK,PanzerDofMgr,PanzerMiniEM,Phalanx,Piro,ROL,Sacado,SEACAS,SEACASAlgebra,SEACASAprepro,SEACASBlot,SEACASConjoin,SEACASEpu,SEACASEx1ex2v2,SEACASEx2ex1v2,SEACASExo2mat,SEACASExo_format,SEACASExodiff,SEACASExomatlab,SEACASExotxt,SEACASExplore,SEACASFastq,SEACASGen3D,SEACASGenshell,SEACASGjoin,SEACASGrepos,SEACASMapvar-kd,SEACASMapvar,SEACASMat2exo,SEACASNemslice,SEACASNemspread,SEACASNumbers,SEACASTxtexo,SEACASAprepro_lib,SEACASChaco,SEACASExoIIv2for32,SEACASExodus,SEACASExodus_for,SEACASIoss,SEACASMapvarlib,SEACASNemesis,SEACASPLT,SEACASSupes,SEACASSuplib,SEACASSuplibC,SEACASSuplibCpp,SEACASSVDI,ShyLU_DDFROSch,ShyLU_NodeBasker,Stokhos,Stratimikos,Tempus,TeuchosComm,TpetraClassic,TpetraCore,TrilinosCouplings,Xpetra,Zoltan2'

D) Filter list of changed pacakge to get only the PT set of packages

CHANGED_PACKAGES_PT_LIST='ALL_PACKAGES,TrilinosFrameworkTests,Amesos2,Anasazi,Belos,Galeri,Ifpack2,Kokkos,MueLu,PanzerAdaptersSTK,PanzerDofMgr,PanzerMiniEM,Phalanx,Piro,ROL,Sacado,SEACAS,SEACASAlgebra,SEACASAprepro,SEACASConjoin,SEACASEpu,SEACASEx1ex2v2,SEACASEx2ex1v2,SEACASExo_format,SEACASExodiff,SEACASExotxt,SEACASExplore,SEACASGen3D,SEACASGenshell,SEACASGjoin,SEACASGrepos,SEACASMapvar-kd,SEACASMapvar,SEACASNemslice,SEACASNemspread,SEACASNumbers,SEACASTxtexo,SEACASAprepro_lib,SEACASChaco,SEACASExoIIv2for32,SEACASExodus,SEACASExodus_for,SEACASIoss,SEACASMapvarlib,SEACASNemesis,SEACASSupes,SEACASSuplib,SEACASSuplibC,SEACASSuplibCpp,ShyLU_DDFROSch,Stokhos,Stratimikos,Tempus,TeuchosComm,TpetraClassic,TpetraCore,TrilinosCouplings,Xpetra,Zoltan2'

E) Generate the *.cmake enables file


real    0m14.791s
user    0m14.710s
sys     0m0.086s
```

That git range diff gives a list of 4342 files shown by the generated file:

```
$ wc -l changed-files.txt 
4342 changed-files.txt
```

This script also generated the file:

```
$ wc -l TrilinosPackageDependencies.xml 
2270 TrilinosPackageDependencies.xml
```

which gives the Trilinos project dependencies.

The generated file `enablePackages.cmake` (the name is whatever you want it to be) looks like:

```
MACRO(PR_ENABLE_BOOL  VAR_NAME  VAR_VAL)
  MESSAGE("-- Setting ${VAR_NAME} = ${VAR_VAL}")
  SET(${VAR_NAME} ${VAR_VAL} CACHE BOOL "Set in enablePackages.cmake")
ENDMACRO()

PR_ENABLE_BOOL(Trilinos_ENABLE_ALL_PACKAGES ON)
PR_ENABLE_BOOL(Trilinos_ENABLE_TrilinosFrameworkTests ON)
PR_ENABLE_BOOL(Trilinos_ENABLE_Amesos2 ON)
PR_ENABLE_BOOL(Trilinos_ENABLE_Anasazi ON)
PR_ENABLE_BOOL(Trilinos_ENABLE_Belos ON)
PR_ENABLE_BOOL(Trilinos_ENABLE_Galeri ON)
PR_ENABLE_BOOL(Trilinos_ENABLE_Ifpack2 ON)
PR_ENABLE_BOOL(Trilinos_ENABLE_Kokkos ON)
PR_ENABLE_BOOL(Trilinos_ENABLE_MueLu ON)
...
```

When there are no Trilinos packages enabled, like when only changing files under `cmake/std/atdm/` with a recent PR merged just today #3158, the script gives:

```
$ time ~/Trilinos.base/Trilinos/commonTools/framework/get-changed-trilinos-packages.sh  be669e4^  be669e4  enablePackages.cmake

***
*** Generating set of Trilinos enables given modified packages from
**** git commit be669e4^ to be669e4
***

TRILINOS_DIR=/home/rabartl/Trilinos.base/Trilinos/commonTools/framework/../..
TRIBITS_DIR=/home/rabartl/Trilinos.base/Trilinos/commonTools/framework/../../cmake/tribits

A) Generate the TrilinosPackageDependencies.xml file


B) Get the set of changed files


Current directory: /home/rabartl/Trilinos.base/Trilinos

Writing file /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/MPI_RELEASE_DEBUG_SHARED/changed-files.txt ...

git diff --name-only be669e4^..be669e4 > /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/MPI_RELEASE_DEBUG_SHARED/changed-files.txt

Current directory: /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/MPI_RELEASE_DEBUG_SHARED

C) Get the unfiltered list of changed Trilinos packages (including 'ALL_PACKAGES')

CHANGED_PACKAGES_FULL_LIST=''

D) Filter list of changed pacakge to get only the PT set of packages

CHANGED_PACKAGES_PT_LIST=''

E) Generate the *.cmake enables file


real    0m0.731s
user    0m0.664s
sys     0m0.065s
```

with the generated file `enablePackage.cmake`:

```
MACRO(PR_ENABLE_BOOL  VAR_NAME  VAR_VAL)
  MESSAGE("-- Setting ${VAR_NAME} = ${VAR_VAL}")
  SET(${VAR_NAME} ${VAR_VAL} CACHE BOOL "Set in enablePackages.cmake")
ENDMACRO()

MESSAGE("-- NOTE: No packages are being enabled!")
```

Note that you can run these generated `*.cmake` files like:

```
$ cmake -P enablePackages.cmake 
-- NOTE: No packages are being enabled!
```

(This is used in the automated testing.)

## Checklist

- [x] My commit messages mention the appropriate GitHub issue numbers.

### CODE REVIEW

**Comment by bartlettroscoe** [comment]
@trilinos/framework,

The [above](https://github.com/trilinos/Trilinos/pull/3134#issuecomment-405696331) PR testing iteration says the `Trilinos_pullrequest_gcc_4.9.3` build failed.  The [CDash results for that PR iteration](https://testing-vm.sandia.gov/cdash/index.php?project=Trilinos&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=PR-3134-test&field2=buildstarttime&compare2=84&value2=NOW) don't show any entry for that build.

Was this another case of the git pull failing?  Is there a way to make this system more robust?

**Comment by bartlettroscoe** [comment]
@william76, as we discussed at our meeting today, I pushed the refactored script `commonTools/test/utilities/get-changed-trilinos-packages.sh` which takes care of generating the `packageEnables.cmake` file to be used for the auto PR tester (or for other purposes).  See the comments ad the top of the script for details.  It should be very easy to use and drop into the PR testing infrastructure to generate the `packageEnables.cmake` file.

**Comment by bartlettroscoe** [comment]
@william76, this passed PR testing.  This needs to be approved again because I pushed an updated commit.  Can you look this over and then approve this PR?   Then we can merge this.

**Comment by william76** [comment]
@bartlettroscoe Reviewed

**Comment by bartlettroscoe** [comment]
@william76, are you ready to start testing and plugging this in to the Trilinos PR tester?  If so, then we should merge this PR branch now.  If not, then I will have an update soon to use more specialized logic for determining when to test all Trilinos packages.

Let me know

**Comment by william76** [comment]
@bartlettroscoe, I'm in ABQ today through Friday and am pretty much stacked up with meetings so I doubt I'll have a ton of time for doing a lot of technical work so we have a couple days.

I'm going to try and meet with @allevin Friday afternoon to get an update on the PR testing system and see if we can get your script integrated in for some manual testing so I'd say we'd want to merge it in by COB Thursday or Friday morning.

**Comment by bartlettroscoe** [comment]
@william76, okay, then let's wait to merge this until I add the specialized logic for Trilinos.  I should have that done by later today and then I will update the PR and we can go from there.

**Comment by jwillenbring** [comment]
@bartlettroscoe To confirm what we talked about briefly and document it here, what you would suggest then if we want to unconditionally exclude a package from the build entire, for example, if a package has not been ported a platform or need not be, you would recommend that we simply add
(using Anasazi as an example)

`-DTrilinos_ENABLE_Anasazi=OFF`

to the list of configuration options for that specific build?

**Comment by bartlettroscoe** [comment]
@jwillenbring, that is one way to do it.   You can also provide another `*.cmake` file called something like `packageExcludes.cmake` that disables these packages as well and include it before `packageEnables.cmake`.   (The first `set(<some-var> <some-value> CACHE <TYPE> ...)` run wins.)

**Comment by bartlettroscoe** [comment]
@jwillenbring,

Or just pass in another argument to the script get-changed-trilinos-packages.sh  for a set of packages to exclude from testing and then filter those out when creating the `packageEnables.cmake` file.  That is likely a better more direct solution.  There are lots of way to skin this cat.

**Comment by bartlettroscoe** [comment]
@jwillenbring, @william76 ,

FYI: I just pushed an updated branch with an updated script, now under `Trilinos/commonTools/framework/get-changed-trilinos-packages.sh`.  This now has specialized logic for Trilinos like not enabling all packages when there are changes under `cmake/std/atdm/` or `cmake/ctest/drivers/<some-machine/`.  I added strong automated tests for both the Python module that implements the specialized Trilinos logic, and the entire script `get-changed-trilinos-packages.sh`.

As I discussed with @jwillenbring yesterday, I added a new TriBITS packages TrilinosFrameworkTests which lives in the directory `Trilinos/commonTools/framework/'.  Any changes to this new code will automatically trigger this package to be enabled and test this logic.

I tested this manually as well for several use cases and it seems to do what it should (see the updated "How has this been tested?" section).

Hopefully this will be easy to plug into the Trilinos auto PR tester.

Merging this PR will radically improve the workflow around extending and maintaining the ATDM Trilinos configuration settings so I am eager to get this merged.

**Comment by bartlettroscoe** [comment]
@trilinos/framework,

Why is the PR tester not testing this updated branch?

File: cmake/README.ProjectCiFileChangeLogic.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  commonTools/framework/ProjectCiFileChangeLogic.py
+```
+
+That is because that is under the base directory for the
+`TrilinosFrameworkTests` TriBITS package.  We want changes to the file
+`ProjectCiFileChangeLogic.py` to trigger the enable of the
+TrilinosFrameworkTests package so as to run the unit tests for that module.
+But the file `ProjectCiFileChangeLogic.py` also needs to exist in this this
+directory so that TriBITS tools will find it and use it automatically.  Using
+a symbolici link satisifes both of these requirements.
```
Comment (mhoemmen) [inline cmake/README.ProjectCiFileChangeLogic.md]: Spelling error: "symbolic" (OK to leave for now and fix in next commit)

File: cmake/tribits/ci_support/filter-packages-list.py [hunk 2]
```diff
@@ -46,30 +46,54 @@
...
 #
 
-usageHelp = r"""filter-packages-list --input-packages-list=P1,P2,... --keep-types=T1,T2,...
+usageHelp = \
+r"""filter-packages-list.py --deps-xml-file=<PROJECT_DEPS_FILE> \
+  --input-packages-list=<P1>,<P2>,... --keep-test-test-categories=<T1>,<T2>,...
+
+This script takes in a comma-seprated list of TriBITS package names
+<P1>,<P2>,... and then filters out the package names for packages that don't
+the package test test categories listed in <T1>,<T2>,... which are given the
```
Comment (mhoemmen) [inline cmake/tribits/ci_support/filter-packages-list.py]: "that don't the package test test categories" -- something happened to that sentence; not quite sure what word to fill in there.

Comment (bartlettroscoe) [reply cmake/tribits/ci_support/filter-packages-list.py]: @mhoemmen, I fixed this in TriBITS 'master' as shown in the updated documentation at:

* https://tribits.org/doc/TribitsDevelopersGuide.html#filter-packages-list-py-help

I will update this in the Trilinos copy in a future Trilinos PR.

File: cmake/README.ProjectCiFileChangeLogic.md (hunk 1 above)
Comment (bartlettroscoe) [reply cmake/README.ProjectCiFileChangeLogic.md]: @mhoemmen, I fixed these in PR #3169.  I would have pushed these to this PR but I don't want to risk the PR builds failing and blocking the merge of this PR.

**Comment by bartlettroscoe** [comment]
I am going to go ahead and merge this.  Any issues with this stuff can be fixed in follow-up PRs (e.g. #3169).

**Pull request merged**